### PR TITLE
REST API

### PR DESCRIPTION
**Pull request by padams**
This PR adds adds a proper REST based API facility to OWA.

## Route Registration

API routes are registered in modules using a new `registerRestApiRoute` method like so:
```
$this->registerRestApiRoute( 
    'v1',               // the version of the REST API
    'siteProfile',  // the api action
    'POST',         // the http request method
    'owa_addSiteRestController',         // the class name
    'controllers/addSiteRestController.php'       // the module path to the class file
);
```

## Requests

The REST API endpoint is accessible at: `your.owainstance.com/api/` 

All requests must include an `apiKey` for authentication. 

There is currently no URL rewriting, so GET requests need to specify param as query string params (e.g. `/api/?owa_version=v1&owa_do=siteProfile&owa_apiKey=somekey&owa_domain=http://www.foo.com&owa_name=Foo`)

## Response

The response to the API request is JSON and contains the following properties:

- _requestID_ - a GUID for the request
- _error_ - an array of any error or validation msgs
- _data_ - the data payload. Will be empty on error.
- _httpResponse_ - info about the http request. currently just contains status_code. More thought needs to go into what's useful to put in here.

## Controllers 

REST API controllers can extend regular web-app controllers so long as there is a `$this->setMode('rest_api'); call in it's `init` method. I don't love this but PHP doesn't support dual class inheritance... 

## Views

Controllers should delegate to a new concrete class extended from `owa_restApiView` view in order to properly set the data payload of the API response using it's `setResponseData` method.

## Wiki

I've added a new page to document the new API methods.

## Old API

The goal is to depreciate the old api.php endpoint once all it's methods are migrated to the REST API.

### CODE REVIEW

**Comment by Maaiins** [comment]
> REST API controllers can extend regular web-app controllers so long as there is a $this->setMode('rest_api'); call in it's init` method. I don't love this but PHP doesn't support dual class inheritance...

You could use traits to accomplish this:
https://www.geeksforgeeks.org/multiple-inheritance-in-php/

File: modules/base/module.php [hunk 1]
```diff
@@ -671,6 +677,18 @@ function registerApiMethods() {
                 '',
                 'edit_sites'
         );
+        
+        $this->registerApiMethod('AddSiteProfile',
+                array( $this, 'addSiteProfile'),
+                array(
```
Comment (Maaiins) [inline modules/base/module.php:683]: Shouldn't we switch from the old `array()` syntax to the new shorthand syntax `[]` ?

**Comment by padams** [comment]
The problem with traits is that methods defined in a trait cannot be over written by another trait. For example, I don't think this would work:

- TraitA->foo();
- TraitB->foo();
- class A uses TraitA
- class B extends class A uses TraitB

I think you then have resolve conflicts, no?

TraitB::foo insteadof A;

Seems like a lot of work, but If we could somehow hide it in an abstract class I'd ok using it. However, in order to maintain controller inheritance, trait conflict stuff would have to be in the concrete class... which might be a lot for a developer to wrap their head around...

Thoughts?

File: modules/base/module.php (hunk 1 above)
Comment (padams) [reply modules/base/module.php:683]: yes. Old habits die hard. :)

**Comment by Maaiins** [comment]
True that´s correct. Seems like a complete overhead

**Comment by padams** [comment]
I think there might be a way to avoid the annoying $this->setMode(... from within the controller. Instead I had forgotten that the CLI and API endpoints already set different config values and constants to indicate the mode that OWA should be in to handle the request. I'm going to refactor this a bit more.

**Comment by padams** [comment]
@Maaiins  I moved all the mode complexity out of way so concrete controllers don't have to remember to set their mode. that's now done on the way in to OWA by the endpoint.

Also I made it possible for a CLIs to output response data returned from the main controller. This way there is better feedback on what the CLI did (similar to the REST API) and allows developers use the output in an automation script. The CLI is still a bit chatty so I may consolidate the output more into a single json object like i did with the REST API.

take a look and let me know your thoughts.

**Comment by Maaiins** [comment]
I´m totally fine with it ;-)

**Pull request merged**